File: models/tailscale.go
package models

import "time"

type NetworkLogResponse struct {
	Logs []NetworkLog `json:"logs"`
}

type NetworkLog struct {
	NodeID          string        `json:"nodeId"`
	Logged          time.Time     `json:"logged"`
	Start           time.Time     `json:"start"`
	End             time.Time     `json:"end"`
	VirtualTraffic  []TrafficFlow `json:"virtualTraffic,omitempty"`
	PhysicalTraffic []TrafficFlow `json:"physicalTraffic,omitempty"`
	SubnetTraffic   []TrafficFlow `json:"subnetTraffic,omitempty"`
}

type TrafficFlow struct {
	Proto   int    `json:"proto"`
	Src     string `json:"src"`
	Dst     string `json:"dst"`
	TxPkts  int    `json:"txPkts"`
	TxBytes int    `json:"txBytes"`
	RxPkts  int    `json:"rxPkts"`
	RxBytes int    `json:"rxBytes"`
}

type DevicesResponse struct {
	Devices []Device `json:"devices"`
}

type Device struct {
	ID             string    `json:"id"`
	Name           string    `json:"name"`
	Hostname       string    `json:"hostname"`
	MachineKey     string    `json:"machineKey"`
	NodeKey        string    `json:"nodeKey"`
	User           string    `json:"user"`
	Tags           []string  `json:"tags"`
	Addresses      []string  `json:"addresses"`
	OS             string    `json:"os"`
	LastSeen       time.Time `json:"lastSeen"`
	Created        time.Time `json:"created"`
	Online         bool      `json:"online"`
	ExitNode       bool      `json:"exitNode"`
	ExitNodeOption bool      `json:"exitNodeOption"`
}

type FlowData struct {
	SourceDevice      *Device    `json:"sourceDevice"`
	DestinationDevice *Device    `json:"destinationDevice"`
	SourceIP          string     `json:"sourceIP"`
	DestinationIP     string     `json:"destinationIP"`
	Protocol          string     `json:"protocol"`
	Port              string     `json:"port"`
	TotalBytes        int        `json:"totalBytes"`
	TotalPackets      int        `json:"totalPackets"`
	FlowType          string     `json:"flowType"`
	FlowCount         int        `json:"flowCount"`
	TimeWindow        TimeWindow `json:"timeWindow"`
}

type TimeWindow struct {
	Start time.Time `json:"start"`
	End   time.Time `json:"end"`
}

type NetworkMap struct {
	Devices   []Device   `json:"devices"`
	Flows     []FlowData `json:"flows"`
	TimeRange TimeWindow `json:"timeRange"`
}

File: frontend/src/types/tailscale.ts
// Tailscale Device Types
export interface TailscaleDevice {
  addresses: string[]
  id: string
  user: string
  name: string
  hostname: string
  clientVersion: string
  updateAvailable: boolean
  os: string
  created: string
  lastSeen: string
  keyExpiryDisabled: boolean
  expires: string
  authorized: boolean
  isExternal: boolean
  machineKey: string
  nodeKey: string
  blocksIncomingConnections: boolean
  enabledRoutes: string[]
  advertisedRoutes: string[]
  clientConnectivity: {
    endpoints: string[]
    derp: string
    mappingVariesByDestIP: boolean
    latency: Record<string, number>
    clientSupports: {
      hairPinning: boolean
      ipv6: boolean
      pcp: boolean
      pmp: boolean
      udp: boolean
      upnp: boolean
    }
  }
  tags: string[]
  tailnetLockError?: string
  tailnetLockKey?: string
}

// Network Flow Log Types
export interface NetworkFlowLog {
  logged: string
  nodeId: string
  start: string
  end: string
  virtualTraffic: TrafficFlow[]
  physicalTraffic: TrafficFlow[]
  subnetTraffic: TrafficFlow[]
}

export interface TrafficFlow {
  src: string
  dst: string
  proto: number
  srcPort: number
  dstPort: number
  rxBytes: number
  txBytes: number
  rxPackets: number
  txPackets: number
}

// UI State Types
export interface NetworkTopology {
  nodes: TailscaleDevice[]
  links: TrafficFlow[]
}

// Filter and Search Types
export interface NetworkFilter {
  timeRange: {
    start: Date
    end: Date
    preset?: 'last-hour' | 'last-day' | 'last-week' | 'custom'
  }
  devices: string[]
  protocols: string[]
  ports: number[]
  tags: string[]
  trafficThreshold: {
    min: number
    max: number
    unit: 'B' | 'KB' | 'MB' | 'GB'
  }
  search: string
}

// API Configuration
export interface TailscaleConfig {
  apiKey: string
  tailnet: string
  baseUrl: string
}

// Error Types
export interface ApiError {
  message: string
  status: number
  code?: string
}

// Protocol mapping
export const PROTOCOL_NAMES: Record<number, string> = {
  1: 'ICMP',
  6: 'TCP',
  17: 'UDP',
  58: 'ICMPv6'
}

// Common ports
export const COMMON_PORTS: Record<number, string> = {
  22: 'SSH',
  53: 'DNS',
  80: 'HTTP',
  443: 'HTTPS',
  993: 'IMAPS',
  995: 'POP3S',
  587: 'SMTP',
  25: 'SMTP',
  110: 'POP3',
  143: 'IMAP',
  21: 'FTP',
  23: 'Telnet',
  3389: 'RDP',
  5432: 'PostgreSQL',
  3306: 'MySQL',
  6379: 'Redis',
  27017: 'MongoDB'
}